### ISSUE TITLE
microk8s cluster not starting

### ISSUE THREAD

**Issue by deejay104** [canonical/microk8s#2344]
Hello,

I have a microk8S cluster with 3 Intel NUC and 8 rasppi all running Ubuntu 20.04.
It was working quite well and when I looked today nothing was working. I have seen that on raspberry nodes the kubelite processes were taking 100% cpu. I have tried to restart few of them but with no luck.

At the end I have stopped all microk8s (microk8s stop) and tried to start only one. But it is not starting.

My last option would be to remove microk8s on all node and re-create the cluster, but as it takes quite sometime, I would prefer to have another option. And also understand why it went wrong. As it should be a production cluster soon I prefer it to be stable.

Thanks
Matthieu

[inspection-report-20210616_225327.tar.gz](https://github.com/ubuntu/microk8s/files/6665832/inspection-report-20210616_225327.tar.gz)

**Comment by ktsakalozos** [comment]
Hi @deejay104 

In the kubelite logs on the attached tarball I see this:
```
Jun 16 22:45:53 pi11 microk8s.daemon-kubelite[37255]: W0616 22:45:53.004602   37255 authentication.go:507] AnonymousAuth is not allowed with the AlwaysAllow authorizer. Resetting AnonymousAuth to false. You should use a different authorizer
Jun 16 22:50:53 pi11 microk8s.daemon-kubelite[37255]: Error: context deadline exceeded
``` 
This probably means that the dqlite cluster was not formed in time.

What are the hardware specs of the clustered nodes? We are interested in the network and the local storage on each node. 

Could you share the contents of `/var/snap/microk8s/current/var/kubernetes/backend/cluster.yaml` from all the nodes as well as a  `ls -l /var/snap/microk8s/current/var/kubernetes/backend/` ? It would also help to share the inspection tarball from the rest of the nodes.

@MathieuBordere do you think this could have been caused by the snapshot compression we released yesterday?

**Comment by deejay104** [comment]
Hi
We have 3 Intel NUC with reliable SSD storage and Raspberry 4 with SD card or USB stick. Network is at 1G with a dedicated Netgear switch.
Since the beginning I see that Pi4 have issue with the storage and time to time errors with the etcd. I have disable the swap, setup journald to be volatile and tried to remove as much as possible all disk access. The NUC were added thinking they could run the dqlite exclusively. But I did not find any option to keep the datastore only on this 3 nodes. If that's possible I am pretty sure it would help for the stability of this cluster (even if it is more a workaround than a fix).

See attached the content from all nodes :
[ls.txt](https://github.com/ubuntu/microk8s/files/6667669/ls.txt)
[cluster.yaml.txt](https://github.com/ubuntu/microk8s/files/6667670/cluster.yaml.txt)

**Comment by MathieuBordere** [comment]
Hi @deejay104,

If the inspection report is from the single node that you tried to start, than I think it's normal output we see, there is no majority of the nodes online, so there can be no leader found for the database to operate.

Can you try starting the nodes with `role 0` from the `cluster.yaml` and report back with inspection tarballs please?

@ktsakalozos I hope it's not.

**Comment by deejay104** [comment]
I have restart microk8s on the 3 nodes with role 0.
Here is the inspect from one of them. Tell me if you need the inspect from the others.
FYI, microk8S shows not started, but I can see the 3 Pi connecting each others to port 19001
[inspection-report-20210617_083630.tar.gz](https://github.com/ubuntu/microk8s/files/6667933/inspection-report-20210617_083630.tar.gz)

**Comment by MathieuBordere** [comment]
Yes, please also attach the report for the others.

**Comment by deejay104** [comment]
[inspection-report-20210617_085500.tar.gz](https://github.com/ubuntu/microk8s/files/6668013/inspection-report-20210617_085500.tar.gz)
[inspection-report-20210617_085508.tar.gz](https://github.com/ubuntu/microk8s/files/6668016/inspection-report-20210617_085508.tar.gz)

**Comment by MathieuBordere** [comment]
I see you started `192.168.1.198`, `192.168.1.201` and `192.168.1.211`.

From the cluster.yaml in your previous post I see `192.168.1.198` doens't have `role 0` but `role 1`. Can you start another `role 0` node please like the `192.168.1.204` or have the roles changed in meanwhile?

Are you willing to run a debug snap if we would provide you with one? @ktsakalozos would this be possible? I could prepare a raft, dqlite and go-dqlite branch to debug.

**Comment by deejay104** [comment]
I have started 192.168.0.204, but same result.
Inspection looks the same, I can upload it if you want, but not sure it will be relevant.
I am ok to run anything that can solve this issue.

My last option is to re-install everything, but at this time I am quite sure this will happen again.

**Comment by deejay104** [comment]
Is there a way to limit which are the nodes that are running the sqlite database ?

**Comment by MathieuBordere** [comment]
`dqlite` nodes can be assigned a role, currently there are 3 roles, `voter (0)`, `stand-by (1)` and `spare (2)` (the numbers refer to the number next to `Role` in `cluster.yaml`.
`voter` and `stand-by` nodes replicate the database, `spare` nodes don't replicate the database but they can be promoted to `stand-by` or `voter` do so. The difference between `voter` and `stand-by` is that the `voter` participates in the quorum.

There is currently no way to say that a certain node should never run the database, but there is a `weight` functionality https://github.com/canonical/go-dqlite/blob/e9c337f1af60e344ec728c042fd5903449c7f7d3/client/client.go#L289 that indicates how heavily loaded a node is and if it should be promoted from `stand-by` to `voter` if needed. Preference is given to nodes with a low weight. One could in theory specify high `weight` to nodes that should only become voter as a last resort.

**Comment by deejay104** [comment]
I can test that. Any idea of how to implement that ?

**Comment by ktsakalozos** [comment]
@deejay104, I just assembled a a snap with debugging flags that would allow @MathieuBordere to have a better view on what is happening. Could you switch to that snap?
```
sudo snap refresh microk8s --channel=1.21/stable/debug
```
`1.21/stbale/debug` is a temporary branch that will disappear in a month and you will be forwarded to 1.21/stable.

To change the weight of a node you have to get the node IP and port from `/var/snap/microk8s/current/var/kubernetes/backend/cluster.yaml`. The following command changed the weight of a node `127.0.0.1` listening on port `19001` to `123`.
```
sudo -E /snap/microk8s/current/bin/dqlite -c /var/snap/microk8s/current/var/kubernetes/backend/cluster.crt -k /var/snap/microk8s/current/var/kubernetes/backend/cluster.key -s file:///var/snap/microk8s/current/var/kubernetes/backend/cluster.yaml k8s ".weight 127.0.0.1:19001 123"
```

The node with the lowest weight has higher chances to become the dqlite leader.

**Comment by deejay104** [comment]
Thanks, I am trying the weight option on a 'test' cluster.

About the debug version it maybe be not replicated yet :
```
root@pi31:~# snap refresh microk8s --channel=1.21/stable/debug
error: cannot perform the following tasks:
- Download snap "microk8s" (2283) from channel "1.21/stable/debug" (Get https://canonical-bos01.cdn.snapcraft.io/download-origin/canonical-lgw01/EaXqgt1lyCaxKaQCU349mlodBkDCXRcg_2283.snap?interactive=1&token=1623952800_5f2d2d0689ab9e3a9b692ffb807b512da5f49d40: dial tcp: lookup canonical-bos01.cdn.snapcraft.io: No address associated with hostname)

```

Do you want to install this version on all nodes or just 1 or 2 ?

**Comment by MathieuBordere** [comment]
> Thanks, I am trying the weight option on a 'test' cluster.
> 
> About the debug version it maybe be not replicated yet :
> 
> ```
> root@pi31:~# snap refresh microk8s --channel=1.21/stable/debug
> error: cannot perform the following tasks:
> - Download snap "microk8s" (2283) from channel "1.21/stable/debug" (Get https://canonical-bos01.cdn.snapcraft.io/download-origin/canonical-lgw01/EaXqgt1lyCaxKaQCU349mlodBkDCXRcg_2283.snap?interactive=1&token=1623952800_5f2d2d0689ab9e3a9b692ffb807b512da5f49d40: dial tcp: lookup canonical-bos01.cdn.snapcraft.io: No address associated with hostname)
> ```
> 
> Do you want to install this version on all nodes or just 1 or 2 ?

Preferably on all the nodes or at least on the ones with `role 0` (the voters).

**Comment by deejay104** [comment]
Install is in progress.
What I have seen with the first nodes that have already started (2 at this time) is :
``
connectAttemptAll ErrNoAvailableLeader 1623942156
``
Others are coming.

About the weight, how can I make sure that it was applied ?

**Comment by MathieuBordere** [comment]
That's normal, you have 4 voters, so a quorum needs at least 3 nodes.

**Comment by MathieuBordere** [comment]
`sudo -E /snap/microk8s/current/bin/dqlite -c /var/snap/microk8s/current/var/kubernetes/backend/cluster.crt -k /var/snap/microk8s/current/var/kubernetes/backend/cluster.key -s file:///var/snap/microk8s/current/var/kubernetes/backend/cluster.yaml k8s ".describe 127.0.0.1:19001"` should return the weight.

**Comment by deejay104** [comment]
I have updated most of them, here is an inspect, but it seems they are not able to elect a new leader.
(note that pi11,pi12 and pi13 have been re-configure as I had to have them working)
If that can help, I can re-install the cluster so that we are sure everything start.

[inspection-report-20210617_181352.tar.gz](https://github.com/ubuntu/microk8s/files/6671698/inspection-report-20210617_181352.tar.gz)

**Comment by MathieuBordere** [comment]
@ktsakalozos something wrong with certificates? 
`[GODQL] connectAttemptAll err:write handshake: x509: certificate signed by unknown authority (possibly because of "x509: invalid signature: parent certificate cannot sign this kind of certificate" while trying to verify candidate authority certificate "k8s") 1623945340`

**Comment by deejay104** [comment]
That's the 3 node that I had to re-install.

**Comment by MathieuBordere** [comment]
Next to the certificate error, I see very few `[RAFT]` logs which is somewhat concerning. Can you compress and send the contents of `/var/snap/microk8s/current/var/kubernetes/backend/` to mathieu.bordere@canonical.com? (if it doesn't contain sensitive data)

**Comment by deejay104** [comment]
nodes with role 0 were started.
Here are the inspect from them :
[inspection-report-20210617_192632.tar.gz](https://github.com/ubuntu/microk8s/files/6672113/inspection-report-20210617_192632.tar.gz)
[inspection-report-20210617_192745.tar.gz](https://github.com/ubuntu/microk8s/files/6672114/inspection-report-20210617_192745.tar.gz)
[inspection-report-20210617_192832.tar.gz](https://github.com/ubuntu/microk8s/files/6672112/inspection-report-20210617_192832.tar.gz)

**Comment by MathieuBordere** [comment]
My feeling is that the configuration that is known to raft, is not the same as the one from `cluster.yaml`, will look further into this tomorrow. With some luck you can achieve quorum if you also start `192.168.1.197:19001`. But let me get back to you tomorrow.

**Comment by deejay104** [comment]
Thanks
FYI, 192.168.1.197 is one of the NUC on which I had to reconfigure the cluster. It is not part of this cluster anymore.
As I said, if that's easier, I can just reset this cluster and rebuild it from scratch, their is no data, only pods running. My main concern is more to understand what may have break it and then if possible fix it.

**Comment by MathieuBordere** [comment]
It is indeed probably easier/cleaner to rebuild it. If the data is not important to you it's probably wiser that I spend my time on other issues. Thanks for your input.

**Comment by deejay104** [comment]
Do you want that I keep the debug version ?

**Comment by MathieuBordere** [comment]
It's extremely verbose, you will have a lot of logging, I would take the regular version.

**Comment by deejay104** [comment]
ok, I will give it another try.
I have the cluster re-build on the Intel NUC and is working fine.
I have prepared the Raspberry. What I can see (with nothing running on them) is that the kubelite process is taking between 10 and 30% of cpu and a constant rate of writing.
I have set the weight so that I hope they will not became master.
I have removed as much write to the drive as possible (volatile journald, syslog to only warn and crit)
And I hope that this will work. If you have any other suggestion that could prevent the issue that I had I take it ^o^

Thanks for you help.
